### PR TITLE
Coverage

### PR DESCRIPTION
**Pull request by mstokercricut**

### CODE REVIEW

File: Sources/FishyJoesCore/TranslatedTypes/TranslatedFunction.swift [hunk 1]
```diff
@@ -21,7 +21,7 @@ struct TranslatedFunction: TranslatedType {
 
         self.sourceType = .function(parameters.map(\.sourceType), returnType.sourceType)
         self.neutralName = "Function<ReturnType=\(returnType.neutralName), Params=[\(parameters.map { $0.neutralName }.joined(separator: ", "))]>"
-        self.nodeName = "(\(parameters.enumerated().map { "_\($0.offset): \($0.element.nodeName)" }.joined(separator: ", "))) => \(returnType.nodeName)"
+        self.nodeName = "(\(parameters.enumerated().map { "_\($0.offset): \($0.element.definingTSNamespace.map { "\($0)." } ?? "")\($0.element.nodeName)" }.joined(separator: ", "))) => \(returnType.definingTSNamespace.map { "\($0)." } ?? "")\(returnType.nodeName)"
```
Comment (cobbal) [inline Sources/FishyJoesCore/TranslatedTypes/TranslatedFunction.swift]: Use `.nodeType` instead to get the namespaced version.

File: Sources/FishyJoesExecute/CodeGen.swift [hunk 2]
```diff
@@ -182,7 +182,7 @@ extension CodeGen {
             // Build the FishyJoes internal execution helper product
             try cmd("swift", arguments: ["build"] + (codeCoveragePath == nil ? [] : Platform.coverageFlags) + ["--product", "🐟☕️"]).run()
 
-            // Trampoline into fishy-joes-execution-helper via Sourcery
+            // Prepare to execute the fishy-joes-execution-helper via Sourcery by assembling an environment
```
Comment (cobbal) [inline Sources/FishyJoesExecute/CodeGen.swift]: This is a worse description of the overarching task; it just describes the first step.

File: Sources/FishyJoesExecute/CodeGen.swift [hunk 3]
```diff
@@ -308,80 +307,162 @@ extension CodeGen {
...
-                    // Replace each line containing __MODULE_DEPENDENCY__ with one line for each dependency module
-                    func template(line: String) -> [String] {
-                        let line = line.replacingOccurrences(of: "__MODULE_NAME__", with: config.module)
-                        if line.contains("__MODULE_DEPENDENCY__") {
-                            return (config.requiredModules + ["Runtime"]).map {
-                                line.replacingOccurrences(of: "__MODULE_DEPENDENCY__", with: $0)
-                            }
+
+                    // Install the module and its dependencies
+                    if platform == .wasm {
```
Comment (cobbal) [inline Sources/FishyJoesExecute/CodeGen.swift:363]: I don't like "if a { ... }; if b { ... }; if a { ... }; if b { ... }" thing here. I would prefer to group the platforms together

File: Sources/FishyJoesExecute/CodeGen.swift [hunk 4]
```diff
@@ -391,61 +472,129 @@ extension CodeGen {
...
+                                # We are installed as a published package
+                                package_directory=".."
+                            fi
+
+
+                            """
+
+                            for dependency in dependencies {
+                                postinstall += """
+                                    ln -sf "$(realpath $package_directory/\(dependency.npmPackageName)/\(dependency.nodeLibName))" "\(dependency.compiledLibName)"
```
Comment (cobbal) [inline Sources/FishyJoesExecute/CodeGen.swift]: quote variables in bash

File: Sources/FishyJoesExecute/CodeGen.swift [hunk 5]
```diff
@@ -391,61 +472,129 @@ extension CodeGen {
...
+
+                            try cmd("cat")
+                                .input(postinstall)
+                                .output(overwritingFile: "\(platform.outputDir(config))/postinstall.sh")
+                                .run()
+                            try cmd("chmod", "+x", "\(platform.outputDir(config))/postinstall.sh").run()
+                            package.scripts[default: [:]]["postinstall"] = "./postinstall.sh"
+                        }
+                        let packageJsonPath = "\(platform.outputDir(config))/package.json"
+                        let prettyEncoder = JSONEncoder()
```
Comment (cobbal) [inline Sources/FishyJoesExecute/CodeGen.swift]: I created a PrettyJsonEncoder in utilities.swift. Use it instead

File: Sources/FishyJoesExecute/CodeGen.swift [hunk 6]
```diff
@@ -391,61 +472,129 @@ extension CodeGen {
...
+                            .prettyPrinted,
+                            .withoutEscapingSlashes
+                        ]
+                        try cmd("cat")
+                            .inputJSON(from: package, encoder: prettyEncoder)
+                            .output(overwritingFile: packageJsonPath)
+                            .run()
+
+                        // Be a good unix citizen and terminate with a newline
+                        try cmd("echo")
```
Comment (cobbal) [inline Sources/FishyJoesExecute/CodeGen.swift]: PrettyJSONEncoder makes this unneeded

File: Sources/FishyJoesIotaRuntime/AttributedString/AttributedSubstring+iota.swift [hunk 7]
```diff
@@ -178,8 +178,13 @@ public func AttributedSubstring_iotaEquals(
 ) -> Bool.CType {
     let env = Env(envRef)
     return env.catching(to: _exn) {
+        let lhsString = try AttributedSubstring.peekIota(lhs, env: env)
+        let rhsString = try AttributedSubstring.peekIota(rhs, env: env)
+        // TODO: A bug in AttributedSubstring causes a crash if empty substrings are compared for equality; remove when this is fixed
+        // let equal = lhsString == rhsString
+        let equal = lhsString.characters.isEmpty && rhsString.characters.isEmpty ? true : lhsString == rhsString
```
Comment (cobbal) [inline Sources/FishyJoesIotaRuntime/AttributedString/AttributedSubstring+iota.swift]: The technical term for "? true :" is "or"

File: integration-tests/TestAPI-bindings/node-test/AttributedStrings.test.ts [hunk 8]
```diff
@@ -163,14 +226,32 @@ test('Substring', () => {
     const substring = attributedString.substringForRange(range)
     expect(substring.string).toEqual("ello Olá こんにち")
     expect(substring.base.string).toEqual("Hello Olá こんにちは")
+    // expect(substring.substring.equals(substring))
```
Comment (cobbal) [inline integration-tests/TestAPI-bindings/node-test/AttributedStrings.test.ts]: Are these meant to be commented out?

**Review by cobbal — CHANGES_REQUESTED** [review]

File: Sources/FishyJoesIotaRuntime/AttributedString/AttributedSubstring+iota.swift (hunk 7 above)
Comment (mstokercricut) [reply Sources/FishyJoesIotaRuntime/AttributedString/AttributedSubstring+iota.swift]: The technical term for the whole expression is `==`, but I suppose one less ternary in the world is a net win.

File: integration-tests/TestAPI-bindings/node-test/AttributedStrings.test.ts (hunk 8 above)
Comment (mstokercricut) [reply integration-tests/TestAPI-bindings/node-test/AttributedStrings.test.ts]: 🤦

**Review by cobbal — APPROVED** [review]

**Pull request merged**